### ISSUE TITLE
Unpredictable behavior of grid in rendering graph.

### ISSUE THREAD

**Issue by Eternal127** [facultyai/dash-bootstrap-components#182]
Hi, I am rendering a bar chart within a tab and everything is working fine, although when I tried putting a graph in second tab it shows very strange behavior. 
I have given md=12 and lg=12 to the column within a tab and then a container named body contains these tabs. Now my my graph in first tab is being rendered just fine but the graph in second tab spans the area of only 6 grid columns, and when I click on inspect element in browser it immediately expands to 12 grid columns, and also html contains col-md-12 and col-lg-12.
In nutshell when first the page is loaded graph spans for area of 6 grid columns but after clicking on inspect element it expands to 12 grid column, however html has col-md-12 and col-lg-12 since when page was loaded. 

Here is my code.

````
team_player_graph = dcc.Graph(
        id='team_player_bar',
        figure = {
            'data':[
                go.Bar(
                    x = [],
                    y = [],
                    text = [],
                    textposition = 'auto'
                )
            ],
            'layout': go.Layout(
                title = 'Please select a player and team',
                xaxis = {
                    'title': 'matches',
                    'type': 'category',
                    'tick0': 0,
                },
                yaxis = {'title': 'Runs'}
            )
        }
)

player_vs_team_tab = dbc.Card(
    dbc.CardBody(
        [
            dbc.Row(
            [
                dbc.Col(
                    player_team_dropdown,
                    md=6,
                    lg=6
                ),
                dbc.Col(
                    team_dropdown,
                    md=6,
                    lg=6
                ),
                
            ]
        ),
        dbc.Row(
            dbc.Col(
                team_player_graph,
                md=12,
                lg=12
            )
        )
        ]
    ),
    className="mt-3",
)

tabs = dbc.Tabs(
    [
        dbc.Tab(player_vs_season_tab, label="Player vs Season"),
        dbc.Tab(player_vs_team_tab, label="Player vs Team"),
    ]
)

body = dbc.Container(
    [
        tabs
    ],
    className="mt-4",
)

app.layout = html.Div(children = [
    jumbotron,
    body,
])
````

Why my graph (or tab) is showing this strange behavior?

**Comment by tcbegley** [comment]
Hi @Eternal127 

Check out [this post](https://community.plot.ly/t/graph-in-tabs-not-extending-to-all-screen/21806/2) on the plotly forums for some details on how to resolve this.

In brief, `Tabs` sets `display:none` on tabs that aren't visible, which means that plotly is unable to infer the correct width when rendering the graph. When the tab is made visible, it doesn't trigger a redraw. As you noticed, if the window is resized (for example by opening the inspector) it triggers a redraw and everything is sized correctly.

The best solution for now is to use the `active_tab` prop of `Tabs` to dynamically render content when a different tab is selected. Something like this

```python
app.layout = html.Div([
    dbc.Tabs(
        [
            dbc.Tab(label="Tab 1", tab_id="tab-1"),
            dbc.Tab(label="Tab 2", tab_id="tab-2")
        ],
        id="tabs"
    ),
    html.Div(id="tab-content"),
])


@app.callback(
    Output("tab-content", "children"),
    [Input("tabs", "active_tab")],
)
def tab_content(active_tab):
    if active_tab == "tab-1":
        return tab1
    elif active_tab == "tab-2":
        return tab2
    # can return e.g. error message if neither of the above satisfied

```

If creating your graphs is computationally expensive, I would consider saving the figure objects in a `dcc.Store` on page load, then loading the right one from there each time a different tab is selected.

**Comment by Eternal127** [comment]
Thanks, It works.

**Comment by Eternal127** [comment]
Everything works  as expected but now there is another issue.
I have basically two drop-downs and a graph in both the tabs, now page is displaying drop-downs twice in both the tabs. like
Drop-downs
My graph
again Drop-downs here.

Is the callback firing twice? If yes, then why only drop-down is rendered twice and not graph.
If I comment this callback then everything is rendered only once.
I am not able to get what is going on here.
I will be really helpful if you can guide me.

**Comment by tcbegley** [comment]
Are you passing children to the `Tab` components? If you are using a callback to switch tabs, the `Tab` components themselves should be empty. E.g. 

```python
dbc.Tab(label="Tab 1")
```

or to be really explicit if you want

```python
dbc.Tab(None, label="Tab 1")
```

**Comment by Eternal127** [comment]
yes, you are right. I figured it out after some digging, thank for the help though. I'm loving this library.
